src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java

```diff
@@ -0,0 +1,200 @@
+* 
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+* OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+* HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+* WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
+* THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+package htsjdk.variant.vcf;
+
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import htsjdk.samtools.util.AbstractIterator;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.RuntimeIOException;
+import htsjdk.tribble.readers.AsciiLineReader;
+import htsjdk.tribble.readers.AsciiLineReaderIterator;
+import htsjdk.tribble.readers.PositionalBufferedStream;
+import htsjdk.variant.bcf2.BCF2Codec;
+import htsjdk.variant.bcf2.BCFVersion;
+import htsjdk.variant.variantcontext.VariantContext;
+
+/**
+ * An Class building {@link VCFReaderIterator}
+ * 
+ * Example:
+ * 
+ * <pre>
+ * VCFIterator r = new VCFIteratorBuilder().open(System.in);
+ * while (r.hasNext()) {
+ *     System.out.println(r.next());
+ * }
+ * r.close();
+ * </pre>
+ * 
+ * @author Pierre Lindenbaum / @yokofakun
+ * @see {@link VCFIterator}
+ *
+ */
+
+public class VCFIteratorBuilder {
+
+    /**
+     * creates a VCF iterator from an input stream It detects if the stream is a
+     * BCF stream or a GZipped stream.
+     * 
+     * @param in
+     *            inputstream
+     * @return the VCFIterator
+     * @throws RuntimeIOException
+     */
+    public VCFIterator open(final InputStream in) throws IOException {
+        if (in == null) {
+            throw new IllegalArgumentException("input stream is null");
+            }
+        // try to read the BCF header
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in, 100);
+        bufferedinput.mark(100);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
+
+        if (bcfVersion != null) {
+            //this is BCF
+            return new BCFInputStreamIterator(bufferedinput);
+        } else {
+            //this is VCF or VCF.gz
+            return new VCFReaderIterator(IOUtil.mayBeGZippedInputStream(bufferedinput));
+        }
+    }
+
+    /**
+     * creates a VCF iterator from an URI It detects if the stream is a BCF
+     * stream or a GZipped stream.
+     * 
+     * @param uri
+     *            the URI
+     * @return the VCFIterator
+     * @throws OException
+     */
+    public VCFIterator open(final String uri) throws IOException {
+        if (IOUtil.isUrl(uri)) {
```

Comment:
Sorry, althought I said that I can add the `java.nio.Path` support, I realized that my change will change the behaviour of this method. To still support urls and files, but with the addition of `java.nio.Path` comming from a `String`, I think that here the best approach will be:

```java
 /**
  * creates a VCF iterator from an Path.
  * It detects if the stream is a BCF stream or a GZipped stream.
  * 
  * @param path a uri like string
  * @return the VCFIterator
  * @throws OException
  */
public VCFIterator open(final String path) throws IOException {
    return open(ParsingUtils.openInputStream(path, null));
}
```

This will simplify the code and do not change the behaviour in a later PR.